### PR TITLE
Add 2.0 stability guarantees

### PR DESCRIPTION
**Pull request by brian-brazil**
As discussed generally consider SDs as unstable, as realistically they
are never going to be. Drop the words "experimental/beta" from most
places in the docs, as users are getting the wrong impression from this.

### CODE REVIEW

File: docs/stability.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+* Configuration file format (minus the service discovery remote read/write, see below)
+* Rule/alert file format
+* Console template syntax and semantics
+
+Things considered unstable for 2.x:
+
+* The [`holt_winters` PromQL function](https://github.com/prometheus/prometheus/issues/2458)
+* Remote read, remote write and the remote read endpoint
+* GRPC APIs
+* Flag meaning and names
```
Comment (brancz) [inline docs/stability.md]: This was previously declared stable in 1.x, no? What is our reasoning behind this?

Comment (brian-brazil) [reply docs/stability.md]: Only the names were stable in 1.x, and it caused oddities. This wording enables us to hard break a user if there's a flag change they really need to think about.

Comment (brancz) [reply docs/stability.md]: I'm happy with this decision. Just wanted to clarify.

**Comment by brancz** [comment]
lgtm

File: docs/stability.md (hunk 1 above)
Comment (grobie) [reply docs/stability.md]: If there is a flag change they really have to think about, we have to release the major version. I'm against doing our own versioning but following semver, so that version numbers have some meaning.

Comment (brancz) [reply docs/stability.md]: I'm specifically thinking about `--external-url` and `--route-prefix` as they lead to frequent confusion. If we want to be more detailed, I would like to exclude those from the stability guarantees. I can see how flags, is something that people would likely complain about if they break and not allow the prometheus server to start in a consecutive release.

Comment (brian-brazil) [reply docs/stability.md]: What if it's a minor flag that most users don't touch, but matters in some obscure use case? For example if anyone was using the flags for talking out to influx we silently started dropping their data on the floor at some point during 1.x. That is not good behaviour, breaking them explicitly would have been better than breaking them silently.

**Comment by brian-brazil** [comment]
> I'm against watering down the stability guarantees in that way. It doesn't make sense to pretend we're doing semver if we exlcude so many parts again.

What would you propose instead? Having major releases every month or two? That kinda defeats the purpose of having major releases. We have several features which are not complete yet, and shouldn't hold back the stability guarantees of complete features due to the more experimental stuff.

File: docs/stability.md (hunk 1 above)
Comment (brian-brazil) [reply docs/stability.md]: > I can see how flags, is something that people would likely complain about if they break and not allow the prometheus server to start in a consecutive release.

That is something some users always complain about, as they don't understand why a hard break is a necessary sometimes. They would prefer the binary to start but now do undesirable things - which they'll likely also blame us for once they discover that. There's no way to win for us, and I'd prefer to be explicit about breakage.

**Comment by beorn7** [comment]
It is common that major version numbers go up quickly if using semver. (An internal core library at SoundCloud has just reached 51.9.0…)

> Having major releases every month or two? That kinda defeats the purpose of having major releases.

One might say that optimizing for rare major releases kinda defeats the purpose of semver.

I'm fine with some very reasonable concessions to practicality with semver, but the list of exceptions is now both long and includes broad and potentially very invasive areas (flag meaning and names,
service discovery integrations). Plus, in the end, a wild card _Any feature listed as experimental or subject to change_. If you don't know the rationale behind all of that, it appears even worse.

**Comment by beorn7** [comment]
To clarify: _Any feature listed as experimental or subject to change_ has an overlap with the list item. On its own, it's a fine item. I suggest: _Features listed as experimental or subject to change, which include <short list of experimental features>_.  The only other items on the list should be the more arcane things like storage format upgrades, exported metrics, and UI HTML.

**Comment by brian-brazil** [comment]
> One might say that optimizing for rare major releases kinda defeats the purpose of semver.

We don't follow semver, and this document is about what we do follow.

> The list of exceptions is now both long and includes broad and potentially very invasive areas (flag meaning and names,
service discovery integrations).

This isn't a big change. Compared to 1.x, the only change is that ec2 SD goes back to unstable and the flag names.
It is a longer list than I'd like, but the bottom 6-7 are going to be there long term I expect.

The only one I'm worried about is that per discussions earlier today that we may need to mark the whole v2 api as unstable (not just the gRPC parts), which I'm not exactly happy about.

> Any feature listed as experimental or subject to change has an overlap with the list item.

The rest of my PR removes those overlaps in other parts of those docs, so I'm not quite getting what you're suggesting.

Can you clarify? Do you mean to list that experimental API explicitly both here and in the API docs, to change the header on the list, or something else? Maybe move it out of the list to a separate bit of text?

> If you don't know the rationale behind all of that, it appears even worse.

The first and last paragraphs were intended to mitigate that. Do you think they should be expanded with more of an explanation?

**Comment by grobie** [comment]
> We don't follow semver, and this document is about what we do follow.

You're contradicting what we agreed to as a project and wrote down for the Prometheus 1.x. https://docs.google.com/document/d/1TFhL6kEasDxkk3pxsHcM6TqSx4E4NI3X2kshNNQ6-X0/edit#heading=h.h98dvniv1d9k

If you want to get rid of semantic versioning, please start a thread to prometheus-developers@ but don't sneak this in into a docs PR.

**Comment by brian-brazil** [comment]
> If you want to get rid of semantic versioning, please start a thread to prometheus-developers@ but don't sneak this in into a docs PR.

I'm not proposing to change our approach to semantic versioning, what's in this PR is the same general approach as in that doc and as we used for 1.x. Prometheus does not use strict semantic versioning.

**Comment by grobie** [comment]
> So far it was applied sporadically, in the future it must be applied strictly – no exceptions.

That's what was agreed upon. Your change is adding even more exceptions. :-1: to include flags. This is a breaking change and requires a new major release version.

**Comment by brian-brazil** [comment]
> That's what was agreed upon. 

I'd suggest reading the rest of of the document, strict semver was not what was agreed or ultimately published at https://prometheus.io/blog/2016/07/18/prometheus-1-0-released/.

> This is a breaking change and requires a new major release version.

We just had 2.0, this is us catching up on the docs for 2.0.

As is stands 2.x technically has no stability guarantees, as our currently published guarantees to users only extend to 1.x.

File: docs/stability.md (hunk 1 above)
Comment (grobie) [reply docs/stability.md]: So what flags are you worried about. The influx flags were related to a feature which was experimental to begin with. The only change which came up was related to proxy behavior. I believe if we come up with a better implementation, it should be done with a new flag name so existing documentation and blog posts can't contradict the new behavior. That can be introduced without changing existing flag behavior.

Comment (brian-brazil) [reply docs/stability.md]: There's no particular ones concerning me currently, but I'd like to keep our options open. For example I can imagine issues arising around the block durations, which are only theoretically there for testing.

Comment (grobie) [reply docs/stability.md]: So there doesn't exist a reason to exclude all flags from the stability guarantees. The only time it caused an issue was solely for an experimental feature which was already excluded from the stability guarantees any way. Given the tsdb flags are not meant for production usage anyway, I'd be fine explicitely excluding these as you seem to prefer. For anything else, we'd need to bump the major version.

You might have a different reading, but I find the linked document we agreed upon quite clear. The goal is not to follow the semver spec less with more releases, but the opposite. I consider it vastly more important to have a semantically meaningful major version, than to keep options open for hypothetically cases and being afraid of major release bumps. To quote the "actual versioning" section of the document a last time.

> To stay relevant we should not be afraid of breaking. Releasing major versions frequently will help with that.

Either reduce flag changes to the dev related tsdb options, or start a thread on prometheus-dev about your opinion on versioning. I don't have anything else to add here anymore.

Comment (brian-brazil) [reply docs/stability.md]: > So there doesn't exist a reason to exclude all flags from the stability guarantees. 

There is no glaring reason right now today, this is about giving us flexibility for the future which we cannot predict. 2.x will be around for a while, and more flags will be added.

> The only time it caused an issue was solely for an experimental feature which was already excluded from the stability guarantees any way.

Given our wording, it was not permitted to hard break those users - even for an experimental feature - as we guaranteed a Prometheus would always start with old flags.
Just because a feature is marked experimental doesn't mean we should let users silently lose data if we can avoid it.

Permitting silent breakage but not explicit breaking is bizarre to me. @juliusv also previously expressed concern when the implications of the 1.x wording became clear.

> I find the linked document we agreed upon quite clear. 

We never agreed the whole doc to my recollection. It was a focus for discussion that petered out, as typically happens with these sorts of doc.

> I consider it vastly more important to have a semantically meaningful major version, than to keep options open for hypothetically cases and being afraid of major release bumps.

That's your opinion. On the basis that the 6 minor releases of 1.x with breaking changes were minor rather than major, that is not the standing policy of the project. (And all 8 minor releases were breaking by a strict definition).

So while the discussion doc may say "Releasing major versions frequently will help with that.", that's not at all what we did in practice.

> or start a thread on prometheus-dev about your opinion on versioning.

If you believe that we should go full semver then the onus is on you to gain consensus on that, as that is not the effective standing policy.

**Comment by beorn7** [comment]
https://prometheus.io/docs/introduction/faq/#how-stable-are-prometheus-features-storage-formats-and-apis is what we have communicated to our users. Obviously, we haven't follow semver strictly, and never will. It's also hardly possible to do so in a provable way as one person's bugfix is another person's breaking change (obligatory reference: https://xkcd.com/1172/ ). However, the alternative to “strict adherence” is not “no adherence”. There is such thing as following semver in spirit with best intentions. I believe we followed it in spirit, and none of our 1.x minor releases was intentionally breaking for users of non-experimental features (if we exclude the XKCD-style breakage).

Declaring broad areas relevant for every user as outside of stability guarantees is against the spirit of semver as applied so far.

**Comment by brian-brazil** [comment]
> I'd suggest to list those bullet points as sub-points underneath something like

I see what you're proposing now, that makes sense. I've moved temporary items under there.

> If I don't know that Prometheus will rarely change the flags I'm most likely using, I'd be very scared and won't put a lot of meaning into the version numbers at all.

This was the case with 1.x, as flag meanings were allowed to change.

> Declaring broad areas relevant for every user as outside of stability guarantees is against the spirit of semver as applied so far.

I agree with you.

What I want to do is to avoid the weird situation the exact wording of the 1.x guarantees got us into. It lead to perverse incentives where silent breakage and data loss was the result rather than hard breakage. That's what I'm trying to avoid happening again.

I believe that either flags should be fully covered by stability guarantees, or they aren't - not anywhere in between. At our current maturity I don't think they can realistically be stable, as that prevents us from doing things like the target heap size improvements you did. Which was a nice graceful change for end users flag-wise.

I'm not saying we should break major flags like the data storage directory, I'm saying we should give ourselves manoeuvring room for similar beneficial changes that touch on flags in future.

**Comment by beorn7** [comment]
I think we are once more at a point where I understand what your rationale is and what the good intentions behind your suggestions are, but where I still do not agree it's the right way and where I'm not sure you understand why I think so.

The way we handled the heap-size flag was exactly in the spirit we (probably) both wanted: The old flags stayed in place, and a similar behavior was emulated so that most users could probably have ignored the change if they had had no need to benefit from the new feature. While a blank declaration of "flags are outside of stability guarantees" certainly allows us to handle things that way, we have in fact handled them under much stricter guidelines.

We are not in a situation where we accept any legal obligation. Thus, we don't have to secure ourselves against lawyers who will sue us for formal violations of any guarantee. Thus, we don't have to harm our users by issuing formally correct but practically worthless guarantees. Since flags are of central operational importance, by excluding flags from stability guarantees (and hence semver semantics), we are creating a situation where the semver semantics is meaningless for each of our users. Only the changelog is relevant, which has to be studied and understood carefully, and each upgrade has to be handled as a (semver-style) major release upgrade.

The message I want to send to our users is very different from ”flags are excluded”: “As long as you aren't using any features marked as experimental/unstable, a minor release upgrade can be performed without any operational adjustments and still very little risk that anything will break.” Obviously, this is not rocket science, and things can always go wrong in unexpected ways. But that's true even with bugfix releases, see https://xkcd.com/1172/ .

**Comment by brian-brazil** [comment]
> I'm not sure you understand why I think so.

I understand where you're coming from, I don't think it's practical to go about things in that exact way.

I think we agree on the goal and what the actual end result for users should be, I think the disagreement is around how we go about saying it. How would you propose to word this around flags?

I suspect getting the right level of nuance and flexibility in would be too wordy, and would worry users more than the wording I'm proposing here. Even though the actual guarantee is stronger as you say.

> “As long as you aren't using any features marked as experimental/unstable, a minor release upgrade can be performed without any operational adjustments and still very little risk that anything will break.”

I've tweaked that and used it instead of the sentence I had.

**Comment by brian-brazil** [comment]
Here's an idea: how about not mentioning flags at all on this page? That'd leave it to the prose to imply our stance.

**Comment by brian-brazil** [comment]
In relation to v2/admin APIs, I've talked with @Gouthamve and @fabxc and what we're doing is adding the new admin apis to v1 and marking all of v2 unstable. This means that we'll only have stable admin apis from Prometheus 2.1, but at least we'll have them.

File: docs/stability.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+---
+title: API Stability
+sort_rank: 8
+---
+
+# API Stability Guarantees
+
+Prometheus promises API stability within a major version, and strives to avoid
+breaking changes for key features. Some features which are cosmetic, still
```
Comment (beorn7) [inline docs/stability.md]: Comma before "which".

Comment (beorn7) [reply docs/stability.md]: "which" starts a non-restrictive relative clause, which (according to most systems of English punctuation) need to be enclosed in commas. This, BTW, means we also need a comma after "services", which I missed to mention. Perhaps that clarifies things.

On the other hand, we don't need to invest time into this. If it is not clear to you at this point, either trust me or leave it. No strong feelings from my side about it, it's just a nit.

Comment (brian-brazil) [reply docs/stability.md]: I see what you mean now, changed.

**Comment by brian-brazil** [comment]
Changing to be against master (rather than release-2.0) so this will be included in 2.1.
